src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java

```diff
@@ -0,0 +1,881 @@
+     * @return never {@code null}.
+     */
+    private static String composeColumnSummaryHeader(final String commandLine, final GroupBy groupBy,
+                                                     final int exonCount, final long exomeSize) {
+        return String.format(
+                String.join("\n",
+                    "##fileFormat  = tsv",
+                    "##commandLine = %s",
+                    "##title       = Summary counts per %s",
+                    "##metaData    = {",
+                    "##    exonCount = %d,",
+                    "##    exomeSize = %d (bp)",
+                    "##}",
+                    String.join("\t","GROUP","SUM","AVG.BP")),
+                commandLine, groupBy.friendlyName(),exonCount, exomeSize);
+    }
+
+    /**
+     * Composes the row summary output header.
+     * @param commandLine the execution command line.
+     * @param groupBy the value of the group-by argument used.
+     * @param columnCount number of count-column involved in the analysis.
+     * @return never {@code null}.
+     */
+    private static String composeRowOutputHeader(final String commandLine, final GroupBy groupBy,
+                                                       final int columnCount) {
+        return String.format(
+                String.join("\n",
+                    "##fileFormat  = tsv",
+                    "##commandLine = %s",
+                    "##title       = Summary counts per exon",
+                    "##metaData = {",
+                    "##    groupBy     = %s,",
+                    "##    columnCount = %d",
+                    "##}",
+                    String.join("\t","CHROM","START","STOP","SUM","AVG.COL.BP")),
+                commandLine,groupBy,columnCount);
+    }
+
+    /**
+     * Composes the main output header.
+     *
+     * @param commandLine the tool command line.
+     * @param groupBy the group-by argument used.
+     * @param countColumnNames the column names.
+     * @return never {@code null}.
+     */
+    private static String composeMatrixOutputHeader(final String commandLine, final GroupBy groupBy,
+                                                          final List<String> countColumnNames) {
+        final String countColumnHeaderString = String.join("\t",countColumnNames).replace("%","%%");
+        final String formatString = String.join("\n",
+                "##fileFormat  = tsv",
+                "##commandLine = %s",
+                "##title       = Read counts per exon and %s",
+                String.join("\t","CHROM","START","STOP",countColumnHeaderString));
+        return String.format(formatString,commandLine,groupBy.friendlyName());
+    }
+
+    /**
+     * Count column header name substitution patterns.
+     *
+     * Each element of this enum represent a different substitution pattern and
+     * provides operations to apply it to a given count column name string.
+     */
+    protected enum CountColumnNamePattern {
```

Comment:
hmm this logic of three different ways of grouping is now duplicated to this enum (so we have 2 enums and 1 real class hierarchy). Can this all be collected in the class hierarchy?